File: mathewDavidov/assignments/mergeIntervals/lc56/lc56.java

/**
 * Problem Statement:
 * Given a list of intervals, merge all the overlapping intervals to 
 * produce a list that has only mutually exclusive intervals.
 * 
 * Example 1:
 * Input: intervals = [[1,3],[2,6],[8,10],[15,18]]
 * Output: [[1,6],[8,10],[15,18]]
 * Explanation: Since intervals [1,3] and [2,6] overlaps, merge them into [1,6].
 * 
 * Edge cases:
 * Null / empty array -> return
 * Array with one intervals -> return interval as it is
 * 
 * sort the 2D array based on first index (start)
 * initialize start at the first interval 0th index
 * initialize end at the first interval 1st index
 * for every interval from 1 to end:
 *   if the start of the current interval is less than or equal to the end:
 *     update the end to the max of the interval end and the end
 *   otherwise:
 *     push a new interval containing the start and end
 *     update the start and end to the current interval's start and end
 * push the interval of the start and end
 * return the resulting merged intervals
 * 
 * Runtime:
 * Time: O(n * log n)
 * Space: O(n)
 */
import java.util.*;

public class lc56 {
    public static int[][] merge(int[][] intervals) {
        if (intervals == null || intervals.length < 2) {
            return intervals;
        }

        List<int[]> mergedIntervals = new ArrayList<>();

        Arrays.sort(intervals, (i1, i2) -> i1[0] - i2[0]);

        int start = intervals[0][0];
        int end = intervals[0][1];

        for (int i = 1; i < intervals.length; i++) {
            if (intervals[i][0] <= end) {
                end = Math.max(end, intervals[i][1]);
            } else {
                mergedIntervals.add(new int[] { start, end });
                start = intervals[i][0];
                end = intervals[i][1];
            }
        }

        mergedIntervals.add(new int[] { start, end });

        // Java's toArray(T[] a) requires a runtime type of the resulting array type
        return mergedIntervals.toArray(new int[0][]);
    }

    public static void main(String[] args) {
        int[][] input = new int[][] { { 1, 3 }, { 2, 5 }, { 7, 9 } };
        System.out.println("Merged intervals: " + Arrays.deepToString(merge(input)));

        input = new int[][] { { 6, 7 }, { 2, 4 }, { 5, 9 } };
        System.out.println("Merged intervals: " + Arrays.deepToString(merge(input)));

        input = new int[][] { { 1, 4 }, { 2, 6 }, { 3, 5 } };
        System.out.println("Merged intervals: " + Arrays.deepToString(merge(input)));
    }
}

File: mathewDavidov/assignments/mergeIntervals/lc56/lc56.py
# Problem Statement #

# Given a list of intervals, merge all the overlapping intervals to produce a list that has only mutually exclusive intervals.

from __future__ import print_function


class Interval:
  def __init__(self, start, end):
    self.start = start
    self.end = end

  def print_interval(self):
    print("[" + str(self.start) + ", " + str(self.end) + "]", end='')


def merge(intervals):
  merged = []
  # TODO: Write your code here
  return merged


def main():
  print("Merged intervals: ", end='')
  for i in merge([Interval(1, 4), Interval(2, 5), Interval(7, 9)]):
    i.print_interval()
  print()

  print("Merged intervals: ", end='')
  for i in merge([Interval(6, 7), Interval(2, 4), Interval(5, 9)]):
    i.print_interval()
  print()

  print("Merged intervals: ", end='')
  for i in merge([Interval(1, 4), Interval(2, 6), Interval(3, 5)]):
    i.print_interval()
  print()

main()




# Solution
# -----
# def merge(intervals):
#   if len(intervals) < 2:
#     return intervals

#   # sort the intervals on the start time
#   intervals.sort(key=lambda x: x.start)

#   mergedIntervals = []
#   start = intervals[0].start
#   end = intervals[0].end
#   for i in range(1, len(intervals)):
#     interval = intervals[i]
#     if interval.start <= end:  # overlapping intervals, adjust the 'end'
#       end = max(interval.end, end)
#     else:  # non-overlapping interval, add the previous internval and reset
#       mergedIntervals.append(Interval(start, end))
#       start = interval.start
#       end = interval.end

#   # add the last interval
#   mergedIntervals.append(Interval(start, end))
#   return mergedIntervals

# -----

# Time complexity #
# The time complexity of the above algorithm is O(N * logN), where ‘N’ is the total number of intervals. We are iterating the intervals only once which will take O(N), in the beginning though, since we need to sort the intervals, our algorithm will take O(N * logN).

# Space complexity #
# The space complexity of the above algorithm will be O(N) as we need to return a list containing all the merged intervals. We will also need O(N) space for sorting. For Java, depending on its version, Collections.sort() either uses Merge sort or Timsort, and both these algorithms need O(N) space. Overall, our algorithm has a space complexity of O(N).
